### PR TITLE
feat: verify api data

### LINKED ISSUE #780: verify API data
As a developer, I would like to run a test to validate a list of endpoints, so that I am confident that nothing important is broken when making changes to the codebase.

It is a part of this issue to figure out the important endpoints.
For model specific endpoints, only focus on the Human-GEM for now.

### PR DESCRIPTION
**Pull request by e0**
**Related issue(s) and PR(s)**  
This PR closes #780.

**Type of change**  

- [x] Other
 
**List of changes made**  

This PR adds jest integration for the `api` as well as a series of tests focusing on verifying data from various endpoints.

**Screenshot of the fix**  
image

**Testing**  

- A rebuild is needed due to new dependencies
- Instructions on how to test
  - Run the following after sourcing 'proj.sh', rebuilding, and starting the project: `ma-exec api yarn test`

**Further comments**  

Currently the tests are relatively basic but ensures some basic checks for most of the endpoints. Please let me know if there is any test that you think we should add.

**Checklist**  

- [x] My code follows the [NBIS style guidelines](https://github.com/NBISweden/development-guidelines)
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have checked so that the same data as before is returned by the API

### CODE REVIEW

**Review by MalinAhlberg — COMMENTED** [review]
This is so nice :star: ! The tests look simple and clean and easy enough to write.

I'm not very familiar with the api itself, so I can't say anything about the coverage. If you think it's a good idea, I could spend some time tomorrow looking at it to see if I can think of more tests to add. Otherwise, I think the PR could be accepted as is.

**Comment by e0** [comment]
> This is so nice ⭐ ! The tests look simple and clean and easy enough to write.
> 
> I'm not very familiar with the api itself, so I can't say anything about the coverage. If you think it's a good idea, I could spend some time tomorrow looking at it to see if I can think of more tests to add. Otherwise, I think the PR could be accepted as is.

Thanks for the nice comment! We can always add more tests to this in the future as well so I don't think you need to burden yourself with exploring the api too much atm.

File: api/test/gemComparison.test.js
Comment (MalinAhlberg) [inline api/test/gemComparison.test.js:26]: maybe I'm missing something here, but I assume `Reaction.unique.length` won't actually check if all elements are unique? the description of the test ("...should have lists of unique reactions and metabolites") seems to indicate that there should be no duplicated elements.

Comment (e0) [reply api/test/gemComparison.test.js:26]: Ah, I could have probably worded the description better. I was looking at this part of the website as I was writing it. Do you think it makes sense in this context?

image

This is the response data:

image

Comment (MalinAhlberg) [reply api/test/gemComparison.test.js:26]: ah, yes, now I get it :D 
Maybe a rephrase would prevent future confusion?
`"the comparison details should have lists of reactions and metabolites"`

**Review by MalinAhlberg — APPROVED** [review]
As mentioned before, I think this looks very good. If you come up with some nice rephrasing for the comment I got confused about, that's nice. If you don't, that's also fine with me :)

File: api/test/gemComparison.test.js
Comment (e0) [reply api/test/gemComparison.test.js:26]: Addressed in https://github.com/MetabolicAtlas/MetabolicAtlas/pull/830/commits/26a6f3345c95a9162837579b02a0d04ce81fdfe3

Thanks again!

**Review by nanjiangshu — COMMENTED** [review]
Thanks for the elegantly added test cases @e0, well done!
I think they all look very nice. 
We could probably add more tests to include 
```
/svg-thumbnail 
/hpa 
/dataOverlay
/external-db
```
as well. 

One of the most important tests we are lacking here is to verify if the returned data from the API are the same when we make changes to the API. I think this is one of the major reasons we prioritized testing. However, I don't exactly how to implement it with jest here. Maybe we need to write a testing script for every case or do you have better suggestions?

**Comment by e0** [comment]
> Thanks for the elegantly added test cases @e0, well done! I think they all look very nice. We could probably add more tests to include
> 
> ```
> /svg-thumbnail 
> /hpa 
> /dataOverlay
> /external-db
> ```
> 
> as well.
> 
> One of the most important tests we are lacking here is to verify if the returned data from the API are the same when we make changes to the API. I think this is one of the major reasons we prioritized testing. However, I don't exactly how to implement it with jest here. Maybe we need to write a testing script for every case or do you have better suggestions?

Thanks @nanjiangshu ! Those are endpoints good suggestions. Also good point about more specific verification of the data. I think at one point @inghylt noticed the total number of components were wrong. We could add some hardcoded numbers to verify this but they would need to be updated each time if any version changes. Maybe that is ok though, or do we have any suggestions about if we can retrieve these numbers more dynamically?

**Comment by nanjiangshu** [comment]
For total number of components, one way to verify is to check with the number of reactions or metabolites in the file `reactions.tsv` and `metabolites.tsv` for certain GEMS, such as Human-GEM or Mouse-GEM.

For other neo4j queries, such as searching, I don't think there is any reference value we can compare to.

**Comment by e0** [comment]
> For total number of components, one way to verify is to check with the number of reactions or metabolites in the file `reactions.tsv` and `metabolites.tsv` for certain GEMS, such as Human-GEM or Mouse-GEM.
> 
That’s a really good idea, thanks!

**Comment by e0** [comment]
@nanjiangshu I have added some new tests for the endpoints you mentioned now. Most are still quite basic but still nice to have a sanity check for them.

For the returned data, I focused on the integrated models endpoint and added the numbers (component counts and version numbers) manually for now. I initially wanted to do as you suggested but realized that files like `reactions.tsv` are not directly under the `./api` folder. I think it would be nice to have this as self-contained as possible. The component counts returned from the integrated models endpoint are calculated dynamically now (as opposed to reading direction from the json file before). I also included the version numbers here to make it easier to notice an error for whenever we have newer versions of any model.

Please give this another look when you have time. Thanks again for the great suggestions!

File: api/test/search.test.js
Comment (nanjiangshu) [inline api/test/search.test.js:19]: I don't quite understand why this test is failed. The limit is set to `50` in the `search` function. In addition, would it be better to change `toBe(50)` to `toBeLessThanOrEqual(50)`?

File: api/test/integratedModels.test.js
Comment (nanjiangshu) [inline api/test/integratedModels.test.js:5]: You are right. Files such as `reactions.tsv` does not exists in the `api` container. Sorry that I was not thinking well. I think it is fair enough to verify the data like that.

**Review by nanjiangshu — COMMENTED** [review]
Hi @e0, the news tests you have added looks great! I have just a small comment regarding the testing for search result and otherwise I'm ready to approve this PR.

File: api/test/search.test.js
Comment (e0) [reply api/test/search.test.js:19]: Sorry, I should have pointed this out in the PR description. This is “supposed to” be failing right now. We have an actual bug in search at the moment. For metabolites results, we combine metabolites and compartmentalized metabolites before returning the results, where each node type has the limit of 50.

Comment (nanjiangshu) [reply api/test/search.test.js:19]: I see. Thanks for the explanation! However, I think still `toBeLessThanOrEqual` might be better to interpret the expectation.

Comment (e0) [reply api/test/search.test.js:19]: Ah yes, I see your point. That would match better with the meaning of the test. The downside is that 0 results would also pass. I specifically wanted to test this on something that would push the limit to enforce we get 50 back, but maybe I can rewrite the description better or add a comment.

Comment (nanjiangshu) [reply api/test/search.test.js:19]: Aha. I see what you mean. Metabolite for `H2O` has more than maximum hits. Then we should keep it but make a better description for this test.

Comment (e0) [reply api/test/search.test.js:19]: Added [a commit here](https://github.com/MetabolicAtlas/MetabolicAtlas/pull/830/commits/4b1671d0bb7c53e58ab4fd060e10c87f2e05dd84) to address it. Still sounds a bit awkward I think but do you think it's ok?

Comment (nanjiangshu) [reply api/test/search.test.js:19]: How about "model search to have 50 results per component type if it exceeds the limit 50"?

Comment (e0) [reply api/test/search.test.js:19]: That sounds much better. Added another fix for it now. Thank you!

**Review by nanjiangshu — APPROVED** [review]
Looks great to me now!

**Comment by e0** [comment]
Hi @MalinAhlberg, I sent you a new request to review the PR since quite a few more tests have been added. Please don't stress about reviewing it though! It would be nice to have input from @inghylt as well when she is back to the project next week since this is something she did a lot of research work on testing.

File: api/test/metabolites.test.js
Comment (MalinAhlberg) [inline api/test/metabolites.test.js:10]: Check that all sybsystems have `id` and `name`? Is there any specific subsystem for `MAM01199m` that must be present?

**Review by MalinAhlberg — COMMENTED** [review]
I (still) think this is great, but following on Nanjiang's previous comment, I wonder if it would make sense to make the test more specific and detailed. I suppose this would make it easier for us to catch errors, but of course it's a lot of work.
I added some separate comments with more concrete examples.

**Comment by e0** [comment]
> I (still) think this is great, but following on Nanjiang's previous comment, I wonder if it would make sense to make the test more specific and detailed. I suppose this would make it easier for us to catch errors, but of course it's a lot of work.
> I added some separate comments with more concrete examples.

This is a great point @MalinAhlberg and I should have thought more about it prior to implementing this issue. I really like the specific suggestions/questions you brought up as well.

 It was a bit unclear exactly what the scope of this issue should be. Looking at the state now, we have a series of tests that act as a sanity check to make sure that most of the API endpoints are functional, but not necessarily 100% correct. I think adding more tests to validate the details would be a big improvement but it could take quite sometime to complete that goal. Do you think it would make sense to create a new issue that focuses more on unit testing for the `api`? Then we can write tests to focus more on the `neo4j` queries, similar to the test for search, as opposed to the `http` ones. It would also help us wrap up this current issue.

File: api/test/crossReference.test.js
Comment (e0) [reply api/test/crossReference.test.js:17]: I think it would be possible to make an http request to the url and see if the response is successful. I'm not sure if it would be too beneficial though because it only validates one of many possible URLs we have in the database. The intention here was to make sure that an external reference (to Recon3D as opposed to one of our own cross references like HMR) could have an URL.

File: api/test/hpa.test.js
Comment (e0) [reply api/test/hpa.test.js:10]: I'm not sure about this one. Do you have any input @mihai-sysbio?

File: api/test/metabolites.test.js
Comment (e0) [reply api/test/metabolites.test.js:10]: I think a unit test might be suitable as mentioned in [this comment](https://github.com/MetabolicAtlas/MetabolicAtlas/pull/830#issuecomment-1019984656).

**Comment by MalinAhlberg** [comment]
> This is a great point @MalinAhlberg and I should have thought more about it prior to implementing this issue. I really like the specific suggestions/questions you brought up as well.
> 
> It was a bit unclear exactly what the scope of this issue should be. Looking at the state now, we have a series of tests that act as a sanity check to make sure that most of the API endpoints are functional, but not necessarily 100% correct. I think adding more tests to validate the details would be a big improvement but it could take quite sometime to complete that goal. Do you think it would make sense to create a new issue that focuses more on unit testing for the `api`? Then we can write tests to focus more on the `neo4j` queries, similar to the test for search, as opposed to the `http` ones. It would also help us wrap up this current issue.

Yes, that might be the best way, to open new issues for that and let this PR focus on what's already here :+1: .

**Comment by e0** [comment]
Had a meeting with @inghylt, who brought up some good points about this PR. Similarly to what @MalinAhlberg [mentioned before], it could also be good to validate the counts, for example the number of genes and metabolites returned for Cytosol.

More specifically, @inghylt mentioned the following points that made me change my mind and work a bit more on this PR.

- There is value in validating details for single components since it still helps ensure that we didn't break anything.
- We do not necessarily need unit testing for validating this kind of data.

**Comment by e0** [comment]
Rebased from the dev branch now and force pushed. If you want to test this more, it's probably easiest to replace your local branch with the remote one again. Sorry for the inconvenience.

**Comment by e0** [comment]
Hi all, sorry for the re-requests again! I have added some new tests now for more precise validation of the components. Please have another review at this when you have time.

File: api/test/crossReference.test.js
Comment (nanjiangshu) [inline api/test/crossReference.test.js:11]:
```suggestion
  test("should return a JSON object for a given data source", async () => {
```

**Comment by MalinAhlberg** [comment]
> Looks great with newly added test @e0. I just added a minor comment for the test description. However, it seems linting is not working for the folder `api/test` since only `api/src` was configured, am I correct @MalinAhlberg ? Maybe we can create a new issue to include `api/test` to linting as well.

Good point! If it's not causing too much problems, we could add that already in this PR? Hopefully it would just mean adding the folder to the configs. Adding it as a separate issue sounds too.

File: api/test/crossReference.test.js
Comment (e0) [reply api/test/crossReference.test.js]: Thanks for spotting this! The description was actually pasted from another file (data overlay) and I forgot to change the description. I changed it to a more appropriate one here: https://github.com/MetabolicAtlas/MetabolicAtlas/pull/830/commits/fdafc41c6e54b1379ff62da43a20264cc3b869f2

**Comment by e0** [comment]
Good point about the linting! It took some time to figure out how to integrate jest with eslint properly but it seems to be working now. It would be great if you can have another run at this before we merge this PR. There is a new package added now as well so I think you would need to run `build-stack` and `start-stack` again.

**Review by MalinAhlberg — APPROVED** [review]
I really like the new tests! It's nice with some negative test cases!

File: api/test/compartments.test.js
Comment (inghylt) [inline api/test/compartments.test.js:31]: maybe use `NUCLEUS_INFO.reactionsCount` instead of 238?

Comment (e0) [reply api/test/compartments.test.js]: Good catch! Fixed in: https://github.com/MetabolicAtlas/MetabolicAtlas/pull/830/commits/67eec460463e3d38bda2ae10cff7098a5220397a

**Comment by inghylt** [comment]
Great work with this PR! I'm so excited to see that we finally have some testing in place. Really nice feedback from the other reviewers as well :+1:

**Review by inghylt — APPROVED** [review]

**Review by nanjiangshu — APPROVED** [review]
Looks great now!

**Pull request merged**
